Openshift InvalidFormatException  Cannot deserialize value for Kind Template

Hello,

With the openshift-client library

```
<dependency>
    <groupId>io.fabric8</groupId>
    <artifactId>openshift-client</artifactId>
    <version>5.7.3</version>
</dependency>
```

I try to get a template :

```
openShiftClient.templates()
    .inNamespace("foo")
    .withName("bar")
    .get()
```
The template contains an object with :

```
securityContext:
  runAsUser: ${USERID}
```

And a parameter :

```
parameters:
- name: USERID
  required: true
```

When executing the Java code, the error is thrown

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `long` from String "${USERID}": not a valid `long` value
 at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: io.fabric8.openshift.api.model.TemplateList["items"]->java.util.ArrayList[5]->io.fabric8.openshift.api.model.Template["objects"]->java.util.ArrayList[0]->io.fabric8.kubernetes.api.model.batch.v1.Job["spec"]->io.fabric8.kubernetes.api.model.batch.v1.JobSpec["template"]->io.fabric8.kubernetes.api.model.PodTemplateSpec["spec"]->io.fabric8.kubernetes.api.model.PodSpec["securityContext"]->io.fabric8.kubernetes.api.model.PodSecurityContext["runAsUser"])
```

The runAsUser is effectivly a long in the destination object, but in the template It's a string refered by the parameter

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
@@ -32,7 +32,6 @@ import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinition;
 import io.fabric8.kubernetes.api.model.apiextensions.v1beta1.JSONSchemaProps;
 import io.fabric8.kubernetes.api.model.apps.Deployment;
-import io.fabric8.kubernetes.api.model.batch.v1.Job;
 import io.fabric8.kubernetes.client.CustomResource;
 import io.fabric8.kubernetes.model.annotation.Group;
 import io.fabric8.kubernetes.model.annotation.Version;

--- a/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
+++ b/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
@@ -33,7 +33,7 @@ import com.fasterxml.jackson.core.JsonParser;
 import com.fasterxml.jackson.databind.DeserializationContext;
 import com.fasterxml.jackson.databind.JsonDeserializer;
 import com.fasterxml.jackson.databind.JsonNode;
-
+import com.fasterxml.jackson.databind.exc.InvalidFormatException;
 import io.fabric8.kubernetes.api.KubernetesResourceMappingProvider;
 import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
 import io.fabric8.kubernetes.api.model.HasMetadata;
@@ -49,6 +49,9 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
     private static final String API_VERSION = "apiVersion";
 
     private static final Mapping mapping = new Mapping();
+
+    // template deserialization can fail if unless we use generic resources
+    private static ThreadLocal<Boolean> IN_TEMPLATE = ThreadLocal.withInitial(() -> false);
 
     @Override
     public KubernetesResource deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
@@ -84,7 +87,23 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
             if (resourceType == null) {
               return jp.getCodec().treeToValue(node, GenericKubernetesResource.class);
             } else if (KubernetesResource.class.isAssignableFrom(resourceType)){
-                return jp.getCodec().treeToValue(node, resourceType);
+                boolean inTemplate = IN_TEMPLATE.get();
+                if (!inTemplate && "io.fabric8.openshift.api.model.Template".equals(resourceType.getName())) {
+                    inTemplate = true;
+                    IN_TEMPLATE.set(true);
+                }
+                try {
+                    return jp.getCodec().treeToValue(node, resourceType);
+                } catch (InvalidFormatException e) {
+                    if (!inTemplate) {
+                        throw e;
+                    }
+                    return jp.getCodec().treeToValue(node, GenericKubernetesResource.class);
+                } finally {
+                    if (inTemplate) {
+                        IN_TEMPLATE.remove();
+                    }
+                }
             }
         }
         return null;

--- a/kubernetes-tests/src/test/java/io/fabric8/openshift/client/server/mock/TemplateTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/openshift/client/server/mock/TemplateTest.java
@@ -16,6 +16,7 @@
 package io.fabric8.openshift.client.server.mock;
 
 import io.fabric8.kubernetes.api.model.APIGroupListBuilder;
+import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.KubernetesList;
 import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
@@ -45,6 +46,7 @@ import java.util.List;
 import java.util.Map;
 
 import static java.util.Collections.singletonMap;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
@@ -184,13 +186,22 @@ class TemplateTest {
     map.put("PORT", "8080");
     KubernetesList list = client.templates().withParameters(map).load(getClass().getResourceAsStream("/template-with-number-params.yml")).processLocally(map);
     assertListIsServiceWithPort8080(list);
+    assertThat(list.getItems()).singleElement()
+      .isInstanceOf(Service.class)
+      .extracting("spec.ports").asList().singleElement()
+      .hasFieldOrPropertyWithValue("port", 8080);
   }
 
   @Test
   void shouldGetTemplateWithNumberParameters() {
     OpenShiftClient client = new DefaultOpenShiftClient(new OpenShiftConfigBuilder().withDisableApiGroupCheck(true).build());
     Template template = client.templates().load(getClass().getResourceAsStream("/template-with-number-params.yml")).get();
-    assertTrue(Serialization.asYaml(template.getObjects().get(0)).contains("{PORT}"));
+    assertThat(template)
+      .extracting(Template::getObjects).asList()
+      .singleElement()
+      .isInstanceOf(GenericKubernetesResource.class)
+      .extracting("additionalProperties.spec.ports").asList().singleElement()
+      .hasFieldOrPropertyWithValue("port", "${PORT}");
   }
 
   protected void assertListIsServiceWithPort8080(KubernetesList list) {

--- a/kubernetes-tests/src/test/java/io/fabric8/openshift/client/server/mock/TemplateTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/openshift/client/server/mock/TemplateTest.java
@@ -26,6 +26,7 @@ import io.fabric8.kubernetes.api.model.ServicePort;
 import io.fabric8.kubernetes.api.model.ServiceSpec;
 import io.fabric8.kubernetes.api.model.StatusBuilder;
 import io.fabric8.kubernetes.client.utils.IOHelpers;
+import io.fabric8.kubernetes.client.utils.Serialization;
 import io.fabric8.openshift.api.model.ParameterBuilder;
 import io.fabric8.openshift.api.model.Template;
 import io.fabric8.openshift.api.model.TemplateBuilder;
@@ -183,6 +184,13 @@ class TemplateTest {
     map.put("PORT", "8080");
     KubernetesList list = client.templates().withParameters(map).load(getClass().getResourceAsStream("/template-with-number-params.yml")).processLocally(map);
     assertListIsServiceWithPort8080(list);
+  }
+
+  @Test
+  void shouldGetTemplateWithNumberParameters() {
+    OpenShiftClient client = new DefaultOpenShiftClient(new OpenShiftConfigBuilder().withDisableApiGroupCheck(true).build());
+    Template template = client.templates().load(getClass().getResourceAsStream("/template-with-number-params.yml")).get();
+    assertTrue(Serialization.asYaml(template.getObjects().get(0)).contains("{PORT}"));
   }
 
   protected void assertListIsServiceWithPort8080(KubernetesList list) {

